src/service/request_service.rs

```diff
@@ -1,73 +1,65 @@
 use crate::helpers;
 use crate::prelude::*;
 use crate::router::Router;
+use crate::service::BoxedFutureResult;
 use crate::types::{RequestInfo, RequestMeta};
 use hyper::{body::HttpBody, service::Service, Request, Response};
-use std::future::Future;
 use std::net::SocketAddr;
 use std::pin::Pin;
+use std::sync::Arc;
 use std::task::{Context, Poll};
+use tokio::sync::Mutex;
 
 pub struct RequestService<B, E> {
-    pub(crate) router: *mut Router<B, E>,
+    pub(crate) router: Arc<Mutex<Router<B, E>>>,
```

Comment:
Are there performance losses with switching`*mut` to `Arc<Mutex<..>>`?

Reply:
Yes, because there are now operations involved that have more than zero overhead (whereas an unsafe mutable reference has no overhead).

Those operations are:
1. Atomic increment (calling `.clone()` on `Arc<T>`) and decrement (when an `Arc<T>` goes out of scope)
1. Whatever operations are being done in 1) [`tokio::sync::Mutex::lock`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/sync/mutex.rs#L262) and 2) [the resulting future's `poll` method](https://github.com/tokio-rs/tokio/blob/master/tokio/src/sync/batch_semaphore.rs#L283).

Are there benchmarks I can run to look at any performance hits?

Reply:
Ah, I just found https://github.com/routerify/routerify-benchmark

Reply:
Well, it looks like there's a fairly sizable performance hit:

**`master`**:
```
❯ wrk --latency -t4 -c200 -d8s http://127.0.0.1:8081
Running 8s test @ http://127.0.0.1:8081
  4 threads and 200 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   238.26us  106.47us   2.48ms   80.05%
    Req/Sec   158.85k     3.24k  165.23k    73.12%
  Latency Distribution
     50%  243.00us
     75%  283.00us
     90%  321.00us
     99%  610.00us
  5057899 requests in 8.02s, 559.54MB read
Requests/sec: 630782.79
Transfer/sec:     69.78MB
```

**this PR**:
```
❯ wrk --latency -t4 -c200 -d8s http://127.0.0.1:8081
Running 8s test @ http://127.0.0.1:8081
  4 threads and 200 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.75ms   92.56us   2.93ms   77.21%
    Req/Sec    28.75k     1.16k   32.68k    78.75%
  Latency Distribution
     50%    1.77ms
     75%    1.80ms
     90%    1.83ms
     99%    1.96ms
  915557 requests in 8.02s, 101.28MB read
Requests/sec: 114169.52
Transfer/sec:     12.63MB
```

Reply:
> > Can you help me understand why you used unsafe within the RequestService? Why unsafe { &mut *self.router } and not just wrap Router in an Arc/Mutex?
> 
> Actually, because of the performance. If I have used `Arc`, it will impact the performance very badly when you will have a lot of simultaneous requests, as the `Arc` manages a concurrent counter to keep a track of the current active owners and this counter gets updated only once at a time, so if you got 1k requests, this counter will get updated one by one and other new requests will wait which impacts the performance.
> 
> So, in this case with `Mutex`, you can handle only one SINGLE REQUEST AT A TIME, and Obviously we don't want that, but if we use only the pointer, it just gets copied on every request and no impact on the performance.

[Source](https://github.com/routerify/routerify/issues/18#issuecomment-638784963). I think the unsafe was intentional here for performance reasons.

Reply:
> so if you got 1k requests, this counter will get updated one by one and other new requests will wait which impacts the performance.

To clarify, it's not the `Arc` that's the problem, as the cost of function calls happening while generating a response will pretty much always dwarf the cost of the atomic inc/decref operations, it's the `Mutex` that's problematic here because it serializes requests. For a benchmark that sends a zillion tiny requests, the cost of the Mutex locking will be problematic.

Reply:
I'm not yet convinced that this `unsafe` code is actually safe :stuck_out_tongue:

Reply:
In any case, I'll close out this PR for now.

Reply:
@cpcloud Perhaps you can open an issue discussing whether the usage of unsafe is safe? Rousan has explained before his reasoning for using unsafe, however I'm not familiar enough to comment on the validity and safety of the current design.

Reply:
@cpcloud Perhaps you can open an issue discussing whether the usage of unsafe is safe? @rousan has explained before his reasoning for using unsafe, however I'm not familiar enough to comment on the validity and safety of the current design.